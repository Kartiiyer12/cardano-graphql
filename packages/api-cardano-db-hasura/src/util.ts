import { Config } from './Config'
import fs from 'fs-extra'
import path from 'path'

export async function readSecrets (rootDir: string): Promise<Partial<Config['db']>> {
  return {
    database: (await fs.readFile(path.join(rootDir, 'postgres_db'), 'utf8')).toString(),
    password: (await fs.readFile(path.join(rootDir, 'postgres_password'), 'utf8')).toString(),
    user: (await fs.readFile(path.join(rootDir, 'postgres_user'), 'utf8')).toString()
  }
}
