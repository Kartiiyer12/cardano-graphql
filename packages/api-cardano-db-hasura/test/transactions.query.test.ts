import BigNumber from 'bignumber.js'
import path from 'path'
import { gql } from 'apollo-boost'
import { DocumentNode } from 'graphql'
import util from '@cardano-graphql/util'
import { TestClient } from '@cardano-graphql/util-dev'
import { testClient } from './util'

function loadQueryNode (name: string): Promise<DocumentNode> {
  return util.loadQueryNode(path.resolve(__dirname, '..', 'src', 'example_queries', 'transactions'), name)
}

describe('transactions', () => {
  let client: TestClient
  beforeAll(async () => {
    client = await testClient.testnet()
  })

  it('Returns transactions by hashes', async () => {
    const result = await client.query({
      query: await loadQueryNode('transactionsByHashesOrderByFee'),
      variables: {
        hashes: [
          'd13e153c6df662e565b4d5608d4224de21fb387c5371f499d788e82912069324',
          '657ac5fa926f9f047360cbc908a4f6177f2ac9b52d4254c63dcb1312127217f4'
        ]
      }
    })
    expect(result.data).toMatchSnapshot()
  })

  it('Returns transactions by hashes with scripts', async () => {
    const plutusResult = await client.query({
      query: await loadQueryNode('transactionsByHashesOrderByFee'),
      variables: {
        hashes: [
          '39f065894a51ddf75f58f577e648de5990d9eddf239619e969ebb8aa3a0ea551'
        ]
      }
    })
    const timelockResult = await client.query({
      query: await loadQueryNode('transactionsByHashesOrderByFee'),
      variables: {
        hashes: [
          'c4f2f4ec91d2afe932c022b9f671cdf44ab62c35e9b6e582335ed01c82d167b0'
        ]
      }
    })
    expect({ plutusResult, timelockResult }).toMatchSnapshot()
  })

  it('Can return ordered by block index', async () => {
    const result = await client.query({
      query: await loadQueryNode('orderedTransactionsInBlock'),
      variables: { blockNumber: 3037760 }
    })
    expect(result.data.transactions.length).toBe(22)
    expect(result.data.transactions[0].blockIndex).toBe(0)
    expect(result.data.transactions[1].blockIndex).toBe(1)
    expect(result.data.transactions[2].blockIndex).toBe(2)
    expect(result.data.transactions[7].blockIndex).toBe(7)
  })

  it('returns an empty array when the transactions has no outputs', async () => {
    const result = await client.query({
      query: gql`query transactionWithNoOutputs(
          $hash: Hash32Hex!
      ) {
          transactions(
              where: { hash: { _eq: $hash } },
          ) {
              outputs {
                  address
                  value
              }
              outputs_aggregate {
                  aggregate {
                      count
                  }
              }
              inputs_aggregate {
                  aggregate {
                      count
                  }
              }
              totalOutput
          }
      }`,
      variables: { hash: '0576d6e3fabccabe120f19368bdc0b5a181759f845704e7b5eb01a2bfe948610' }
    })
    expect(result.data.transactions.length).toBe(1)
    expect(result.data.transactions[0].outputs_aggregate.aggregate.count).toBe('0')
    expect(result.data.transactions[0].outputs).toEqual([])
    expect(result.data.transactions[0].totalOutput).toEqual('0')
    expect(result.data.transactions[0].inputs_aggregate.aggregate.count).toBe('1')
  })

  it('Can return aggregated data', async () => {
    const result = await client.query({
      query: await loadQueryNode('aggregateDataWithinTransaction'),
      variables: {
        hashes: [
          'b5aff847f8cefe95f3ae06ae5850d19d33301a5aa3aedc618a3729558403e3db'
        ]
      }
    })
    const { transactions: txs } = result.data
    const outputsPlusFee = new BigNumber(txs[0].outputs_aggregate.aggregate.sum.value).plus(txs[0].fee).toString()
    expect(txs[0].inputs_aggregate.aggregate.sum.value).toEqual(outputsPlusFee)
    expect(result.data).toMatchSnapshot()
  })
  it('Can return filtered aggregated data', async () => {
    const result = await client.query({
      query: await loadQueryNode('filteredAggregateDataWithinTransaction'),
      variables: {
        hash: '2819f6a34f9029251eb91309c85d4b42d7d8dd26ed00ea0693463176bac62c45',
        inputsValueGt: '3842014',
        outputsAddress: 'addr_test1vzhv4acdxm6v00m2h62pllw4l6qtymh3pwkpr9urptp8zxccuawfr'
      }
    })
    expect(result.data).toMatchSnapshot()
  })

  describe('metadata', () => {
    it('JSON object', async () => {
      const result = await client.query({
        query: await loadQueryNode('transactionByIdWithMetadataIfPresent'),
        variables: { hash: '45fff8715b952f04cd84f6235b1a6f22d9437c2d60a699317f08bba60e2965a6' }
      })
      expect(result.data).toMatchSnapshot()
    })

    it('JSON string', async () => {
      const result = await client.query({
        query: await loadQueryNode('transactionByIdWithMetadataIfPresent'),
        variables: { hash: '23ae1816db236baa5e231166040ae5458b25f34bb33812b2754429f122f85a0d' }
      })
      expect(result.data).toBeDefined()
    })
  })

  describe('transactions with tokens', () => {
    it('shows the tokens minted and output', async () => {
      const result = await client.query({
        query: await loadQueryNode('transactionsByHashesWithTokens'),
        variables: { hashes: ['fad76833b223b1ba778c4de016c6224cb56f29f48d286ee84090aa80afec1a58'] }
      })
      expect(result.data).toMatchSnapshot()
    })
  })
})
