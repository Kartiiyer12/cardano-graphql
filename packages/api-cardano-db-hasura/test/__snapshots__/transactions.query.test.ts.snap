// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transactions Can return aggregated data 1`] = `
Object {
  "transactions": Array [
    Object {
      "deposit": 0,
      "fee": 174609,
      "hash": "b5aff847f8cefe95f3ae06ae5850d19d33301a5aa3aedc618a3729558403e3db",
      "inputs_aggregate": Object {
        "aggregate": Object {
          "avg": Object {
            "value": "2375822822",
          },
          "count": "1",
          "max": Object {
            "value": "2375822822",
          },
          "min": Object {
            "value": "2375822822",
          },
          "sum": Object {
            "value": "2375822822",
          },
        },
      },
      "outputs_aggregate": Object {
        "aggregate": Object {
          "avg": Object {
            "value": "1187824106.5",
          },
          "count": "2",
          "max": Object {
            "value": "2372148213",
          },
          "min": Object {
            "value": "3500000",
          },
          "sum": Object {
            "value": "2375648213",
          },
        },
      },
    },
  ],
}
`;

exports[`transactions Can return filtered aggregated data 1`] = `
Object {
  "transactions": Array [
    Object {
      "hash": "2819f6a34f9029251eb91309c85d4b42d7d8dd26ed00ea0693463176bac62c45",
      "outputs_aggregate": Object {
        "aggregate": Object {
          "count": "1",
        },
      },
    },
  ],
}
`;

exports[`transactions Returns transactions by hashes 1`] = `
Object {
  "transactions": Array [
    Object {
      "block": Object {
        "number": null,
      },
      "blockIndex": 0,
      "collateralInputs": Array [],
      "deposit": 0,
      "fee": 0,
      "hash": "d13e153c6df662e565b4d5608d4224de21fb387c5371f499d788e82912069324",
      "inputs": Array [],
      "inputs_aggregate": Object {
        "aggregate": Object {
          "sum": Object {
            "value": null,
          },
        },
      },
      "outputs": Array [
        Object {
          "address": "37btjrVyb4KDHDPBVenqrh8tUTVNYX5ZGwjd4r3svqdnwWicGnMZZV1E7nBJVQsDY69co936H9onHpmA3PYSabYH4ibbULphL1CitDgArH9KknBARc",
          "addressHasScript": false,
          "index": 0,
          "paymentCredential": null,
          "value": "19999999999999",
        },
      ],
      "outputs_aggregate": Object {
        "aggregate": Object {
          "sum": Object {
            "value": "19999999999999",
          },
        },
      },
      "redeemers": Array [],
      "scriptSize": 0,
      "scripts": Array [],
      "size": 0,
      "totalOutput": "19999999999999",
      "validContract": true,
      "withdrawals_aggregate": Object {
        "aggregate": Object {
          "sum": Object {
            "amount": null,
          },
        },
      },
    },
    Object {
      "block": Object {
        "number": 1936322,
      },
      "blockIndex": 2,
      "collateralInputs": Array [],
      "deposit": 0,
      "fee": 179757,
      "hash": "657ac5fa926f9f047360cbc908a4f6177f2ac9b52d4254c63dcb1312127217f4",
      "inputs": Array [
        Object {
          "address": "addr_test1qpjettexkjevymrluxr6lvrrq3f4whdq43g78zk93s27cc3hryxtv2vgjpqnxlhqnvc96le2u7mc8gsasmq34nerfjcqrk4smr",
          "sourceTxHash": "e411a2886de918f7c3b9f9a9a12f32b2c64488b3363291d7146ab21f36a29ae7",
          "sourceTxIndex": 0,
          "value": "10000007624162",
        },
      ],
      "inputs_aggregate": Object {
        "aggregate": Object {
          "sum": Object {
            "value": "10000007624162",
          },
        },
      },
      "outputs": Array [
        Object {
          "address": "addr_test1qpjettexkjevymrluxr6lvrrq3f4whdq43g78zk93s27cc3hryxtv2vgjpqnxlhqnvc96le2u7mc8gsasmq34nerfjcqrk4smr",
          "addressHasScript": false,
          "index": 0,
          "paymentCredential": "6595af26b4b2c26c7fe187afb0630453575da0ac51e38ac58c15ec62",
          "value": "10000876045027",
        },
      ],
      "outputs_aggregate": Object {
        "aggregate": Object {
          "sum": Object {
            "value": "10000876045027",
          },
        },
      },
      "redeemers": Array [],
      "scriptSize": 0,
      "scripts": Array [],
      "size": 367,
      "totalOutput": "10000876045027",
      "validContract": true,
      "withdrawals_aggregate": Object {
        "aggregate": Object {
          "sum": Object {
            "amount": "868600622",
          },
        },
      },
    },
  ],
}
`;

exports[`transactions Returns transactions by hashes with scripts 1`] = `
Object {
  "plutusResult": Object {
    "data": Object {
      "transactions": Array [
        Object {
          "block": Object {
            "number": 2993776,
          },
          "blockIndex": 17,
          "collateralInputs": Array [
            Object {
              "value": "10000000",
            },
          ],
          "deposit": 0,
          "fee": 432963,
          "hash": "39f065894a51ddf75f58f577e648de5990d9eddf239619e969ebb8aa3a0ea551",
          "inputs": Array [
            Object {
              "address": "addr_test1qr8a6n53n63payfj8yk3wyne55drg676ypzf2wtfj4ndymhw05f4lcyppn6jqe2dffwum65fpsv037wsl0luszgvkz2q2z9e85",
              "sourceTxHash": "b3587dba3b74f35025f75fac0232fb9df739722a8b65851d8756914607b47eee",
              "sourceTxIndex": 0,
              "value": "100000000",
            },
          ],
          "inputs_aggregate": Object {
            "aggregate": Object {
              "sum": Object {
                "value": "100000000",
              },
            },
          },
          "outputs": Array [
            Object {
              "address": "addr_test1qzrsw4n8nn4p3gchk796swhjq5hhs94cxcq6d3ee3ufj63lw05f4lcyppn6jqe2dffwum65fpsv037wsl0luszgvkz2qdudpd3",
              "addressHasScript": false,
              "index": 0,
              "paymentCredential": "870756679cea18a317b78ba83af2052f7816b83601a6c7398f132d47",
              "value": "96118837",
            },
            Object {
              "address": "addr_test1qzrsw4n8nn4p3gchk796swhjq5hhs94cxcq6d3ee3ufj63lw05f4lcyppn6jqe2dffwum65fpsv037wsl0luszgvkz2qdudpd3",
              "addressHasScript": false,
              "index": 1,
              "paymentCredential": "870756679cea18a317b78ba83af2052f7816b83601a6c7398f132d47",
              "value": "1724100",
            },
            Object {
              "address": "addr_test1wqr0gpnql52gez5jfgsm28accs8yd6nkwhue2gtggskkr3cw62h5u",
              "addressHasScript": true,
              "index": 2,
              "paymentCredential": "06f40660fd148c8a924a21b51fb8c40e46ea7675f9952168442d61c7",
              "value": "1724100",
            },
          ],
          "outputs_aggregate": Object {
            "aggregate": Object {
              "sum": Object {
                "value": "99567037",
              },
            },
          },
          "redeemers": Array [
            Object {
              "scriptHash": "82f065f4f62986d9add2c9418a559bd88fa26325d7996307a67545dd",
            },
          ],
          "scriptSize": 3602,
          "scripts": Array [
            Object {
              "hash": "82f065f4f62986d9add2c9418a559bd88fa26325d7996307a67545dd",
              "serialisedSize": 3602,
              "type": "plutusV1",
            },
          ],
          "size": 4294,
          "totalOutput": "99567037",
          "validContract": true,
          "withdrawals_aggregate": Object {
            "aggregate": Object {
              "sum": Object {
                "amount": null,
              },
            },
          },
        },
      ],
    },
    "loading": false,
    "networkStatus": 7,
    "stale": false,
  },
  "timelockResult": Object {
    "data": Object {
      "transactions": Array [
        Object {
          "block": Object {
            "number": 2288957,
          },
          "blockIndex": 0,
          "collateralInputs": Array [],
          "deposit": 0,
          "fee": 180637,
          "hash": "c4f2f4ec91d2afe932c022b9f671cdf44ab62c35e9b6e582335ed01c82d167b0",
          "inputs": Array [
            Object {
              "address": "addr_test1vzge9w95jgxlyh0l49fs0ugk736deu8tl5r2fqx4qg05jjqfpu5va",
              "sourceTxHash": "94c10df5fceb3567b9d969319dabe07c8d40cdea11a5169df164571a043ad5b0",
              "sourceTxIndex": 0,
              "value": "1000000000",
            },
          ],
          "inputs_aggregate": Object {
            "aggregate": Object {
              "sum": Object {
                "value": "1000000000",
              },
            },
          },
          "outputs": Array [
            Object {
              "address": "addr_test1vzge9w95jgxlyh0l49fs0ugk736deu8tl5r2fqx4qg05jjqfpu5va",
              "addressHasScript": false,
              "index": 0,
              "paymentCredential": "9192b8b4920df25dffa95307f116f474dcf0ebfd06a480d5021f4948",
              "value": "999819363",
            },
          ],
          "outputs_aggregate": Object {
            "aggregate": Object {
              "sum": Object {
                "value": "999819363",
              },
            },
          },
          "redeemers": Array [],
          "scriptSize": 0,
          "scripts": Array [
            Object {
              "hash": "4b03642a00b1f25ff66508ebdcc400cbcec9b0527bd2c39838f820e0",
              "serialisedSize": null,
              "type": "timelock",
            },
          ],
          "size": 429,
          "totalOutput": "999819363",
          "validContract": true,
          "withdrawals_aggregate": Object {
            "aggregate": Object {
              "sum": Object {
                "amount": null,
              },
            },
          },
        },
      ],
    },
    "loading": false,
    "networkStatus": 7,
    "stale": false,
  },
}
`;

exports[`transactions metadata JSON object 1`] = `
Object {
  "transactions": Array [
    Object {
      "metadata": Array [
        Object {
          "key": "721",
          "value": Object {
            "57cdc42f2041cf67ceea61be6a25e913aff31f7f0bb5e59b8b6ad18b": Object {
              "NEWNFTxx0": Object {
                "image": "ipfs://XXXXYYYYZZZZ",
                "name": "Test NFT #0",
              },
            },
          },
        },
      ],
    },
  ],
}
`;

exports[`transactions transactions with collateral shows the collateral inputs and outputs 1`] = `
Object {
  "transactions": Array [
    Object {
      "collateralInputs": Array [
        Object {
          "address": "addr_test1vzmvs72wnfazvkv5gzjdpltee5rkgng4j9llzd5578m8ydgkp6edr",
          "sourceTransaction": Object {
            "hash": "ada4fb7984252af87faf71de9e94c66f24851b0e6dd509df2d1ea9c795694188",
          },
          "sourceTxHash": "ada4fb7984252af87faf71de9e94c66f24851b0e6dd509df2d1ea9c795694188",
          "sourceTxIndex": 1,
          "tokens": Array [],
          "tokens_aggregate": Object {
            "aggregate": Object {
              "sum": Object {
                "quantity": null,
              },
            },
          },
          "value": "11492321",
        },
      ],
      "collateralInputs_aggregate": Object {
        "aggregate": Object {
          "avg": Object {
            "value": "11492321",
          },
          "count": "1",
          "max": Object {
            "value": "11492321",
          },
          "min": Object {
            "value": "11492321",
          },
          "sum": Object {
            "value": "11492321",
          },
        },
      },
      "collateralOutputs": Array [
        Object {
          "address": "addr_test1vzmvs72wnfazvkv5gzjdpltee5rkgng4j9llzd5578m8ydgkp6edr",
          "addressHasScript": false,
          "datum": null,
          "index": 1,
          "paymentCredential": "b6c8794e9a7a26599440a4d0fd79cd07644d15917ff13694f1f67235",
          "script": null,
          "value": "11232872",
        },
      ],
      "collateralOutputs_aggregate": Object {
        "aggregate": Object {
          "avg": Object {
            "value": "11232872",
          },
          "count": "1",
          "max": Object {
            "value": "11232872",
          },
          "min": Object {
            "value": "11232872",
          },
          "sum": Object {
            "value": "11232872",
          },
        },
      },
      "hash": "3fca8f8be6efba69327388dac1d0b37a72f8803ada94f724aa8bbf17b7e38eee",
    },
  ],
}
`;

exports[`transactions transactions with tokens shows the tokens minted and output 1`] = `
Object {
  "transactions": Array [
    Object {
      "block": Object {
        "number": 2455718,
      },
      "blockIndex": 0,
      "deposit": 0,
      "fee": 181693,
      "hash": "fad76833b223b1ba778c4de016c6224cb56f29f48d286ee84090aa80afec1a58",
      "inputs": Array [
        Object {
          "address": "addr_test1qqzekmfvn8vnphl2ny42pglp7fzmg60zuahscsfwalmz7u6sh5k8hkh7ntx7v92qgeyvtrzd96wp0uf5shvhht7rxxqqnt89ae",
          "sourceTxHash": "18401eeaf66d2c5cf65a13f449f5b22046610bf490e9e54c5253208a13a8b1b8",
          "sourceTxIndex": 0,
          "value": "10000000",
        },
      ],
      "inputs_aggregate": Object {
        "aggregate": Object {
          "sum": Object {
            "value": "10000000",
          },
        },
      },
      "mint": Array [
        Object {
          "asset": Object {
            "assetId": "34250edd1e9836f5378702fbf9416b709bc140e04f668cc3552085184154414441636f696e",
            "assetName": "4154414441636f696e",
            "policyId": "34250edd1e9836f5378702fbf9416b709bc140e04f668cc355208518",
          },
          "quantity": "100",
        },
      ],
      "outputs": Array [
        Object {
          "address": "addr_test1qqzekmfvn8vnphl2ny42pglp7fzmg60zuahscsfwalmz7u6sh5k8hkh7ntx7v92qgeyvtrzd96wp0uf5shvhht7rxxqqnt89ae",
          "index": 0,
          "tokens": Array [
            Object {
              "asset": Object {
                "assetId": "34250edd1e9836f5378702fbf9416b709bc140e04f668cc3552085184154414441636f696e",
                "assetName": "4154414441636f696e",
                "policyId": "34250edd1e9836f5378702fbf9416b709bc140e04f668cc355208518",
              },
              "quantity": "100",
            },
          ],
          "value": "9818307",
        },
      ],
      "outputs_aggregate": Object {
        "aggregate": Object {
          "sum": Object {
            "value": "9818307",
          },
        },
      },
      "size": 449,
      "totalOutput": "9818307",
      "withdrawals_aggregate": Object {
        "aggregate": Object {
          "sum": Object {
            "amount": null,
          },
        },
      },
    },
  ],
}
`;
